include/support.h

```diff
@@ -55,6 +55,13 @@ char * safe_strcpy(char (& dst)[N], const char * src) noexcept {
 	return & dst[0];
 }
 
+template<size_t N>
+char * safe_strcat(char (& dst)[N], const char * src) noexcept {
+	const size_t dst_size = sizeof(dst);
+	strncat(dst, src, dst_size - strnlen(dst, dst_size) - 1);
+	return & dst[0];
+}
```

Comment:
Ahh yes; I should have seen that use of N in `safe_strcpy`.  Thanks!

Interesting; if N is the template, then are these functions generated for each unique-sized stack-array? (I suppose they could be inlined like you mentioned, which is essentially what the code would be without the function).

I've only used types in templates; so using size/length is a very neat use of templates!

Reply:
Yes, they are template specializations and generated array size (in debug builds), and hopefully inlined (in release builds). Some STL templates, that are also specialized this way are `std::array` or `std::get`. They are so small, that compiler shouldn't have a problem with figuring out, that they should be inlined (and `noexcept` is there to help with that, but also as a factual statement ;) ).

Perhaps I should add some hints about inlining (not sure if it would help or make it less likely to be inlined though):
- C++ `inline` means something different than C `inline` (e.g. usage in our header-only libraries is correct, usage in DOSBox code might be pointless)
- I see many places in DOSBox code use `__attribute__((always_inline))` or `__forceinline` (confusingly both defined to INLINE) - perhaps use of one of those would make it more likely for compiler to actually inline a function (I read documentation, that contradicted each other, so I'm not sure which one to use).

I wish C++ had Rust-like [safe macros](https://doc.rust-lang.org/reference/macros.html) or clear and concise definition of [inline attribute](https://doc.rust-lang.org/reference/attributes/codegen.html#the-inline-attribute) :(.

Reply:
Presumably the compiler should be pretty smart about figuring out when inlining will improve performance; and if the functions are obvious enough then we might want to just leave it as-is.